Simplify language chooser

julianwachholz opened the pull request:
---
name: Pull request
about: Submit a pull request for this project
assignees: fabiocaccamo

---

**Describe your changes**
- Allows usage of i18n_patterns with prefix_default_language=False
- Change templatetag from simple tag to an inclusion tag
- Reduces complexity by relying on Django's behavior in the set_language view: it will translate any url passed as 'next'. This behavior has been present since Django 1.9.
- Remove individual forms for each language

Reference: https://github.com/django/django/commit/aa5ab114e34645823b219f9a866fd82fc26b427b


**Related issue**
#327

**Checklist before requesting a review**
- [x] I have performed a self-review of my code.
- [x] I have added tests for the proposed changes.
- [x] I have run the tests and there are not errors.

fabiocaccamo commented:
@julianwachholz thank you very much for this cool PR, I will test an review it as soon as possible!

fabiocaccamo commented:
@julianwachholz sorry for the late answer, I finally had some time for testing this PR.

You made some great optimizations, so it's my intention to merge this PR.

However, the PR doesn't solve #327, if I don't prefix default language (`prefix_default_language=False`), when I switch to the default language using the language chooser, I land to the current url that still includes the default language prefix (`http://localhost:8000/en-us/admin/` instead of `http://localhost:8000/admin/`) *(wrong)* and the resulting page is a 404 error page *(correct)*.

<img width="948" alt="Screenshot 2023-11-01 at 12 40 14" src="https://github.com/fabiocaccamo/django-admin-interface/assets/1035294/38053dd9-627f-4820-9cdb-e3db8f9de2d5">

---

At this point, considering that your refactoring looks right to me, it seems to be more a Django `set_language` bug.

I look forward to your feedback.

fabiocaccamo commented:
@julianwachholz any update?

julianwachholz commented:
> @julianwachholz any update?

I will look at it soon but I'm on vacation this week. 🙂

julianwachholz commented:
@fabiocaccamo after retracing your steps I found the culprit: the `set_language` view needs the `LocaleMiddleware` installed when using it like this. So I added an additional check for it.

fabiocaccamo commented:
@julianwachholz the check you added is very useful indeed, thanks! 

Unfortunately it is not the cause of the problem I reported, because I already have `django.middleware.locale.LocaleMiddleware` in my `settings.MIDDLEWARE` list.

fabiocaccamo reviewed (changes requested):
With the check you introduced, I expect to see all tests to fail because there is not `django.middleware.locale.LocaleMiddleware` in `tests.settings.MIDDLEWARE`, but some tests succeeded anyway.

julianwachholz commented:
Thank you for the demo.zip, @fabiocaccamo. I checked it out and wasn't able to produce an error using it. Both with `prefix_default_language=True` and `prefix_default_language=False`.

The tests for Django <= 3.1 are passing because the package doesn't have a `default_app_config` in `__init__.py` ([automatic AppConfig discovery was added in 3.2](https://docs.djangoproject.com/en/4.2/releases/3.2/#automatic-appconfig-discovery)). So I think this is a different bug related to the Django versions supported by `django-admin-interface`.

Because the language picker is optional, I removed the check and it will now fail silently on misconfiguration and require users to set it up correctly as the documentation describes.
As a small addition I added a runtime warning that only shows up when the language chooser is enabled.

merwok commented on admin_interface/templatetags/admin_interface_tags.py:
```diff
@@ -16,45 +17,43 @@
+        warnings.warn(
+            "Language chooser requires 'django.middleware.locale.LocaleMiddleware' "
+            "in your MIDDLEWARE to work.",
+            stacklevel=1,
```
Isn’t 1 the default value? Is it correct here? (to determine: check what file and line of code is shown as the origin of the warning in the output)

julianwachholz replied:
Yes it's the default. The linter was complaining it wants it explicitly.

merwok replied:
Two notes:
- that sounds like a bad linter, or strange configuration
- the point is that stacklevel=1 is very rarely useful, it’s more common to see stacklevel=2 so that the user code with the problem is shown in the traceback

fabiocaccamo commented:
> However, the PR doesn't solve #327, if I don't prefix default language (`prefix_default_language=False`), when I switch to the default language using the language chooser, I land to the current url that still includes the default language prefix (`http://localhost:8000/en-us/admin/` instead of `http://localhost:8000/admin/`) _(wrong)_ and the resulting page is a 404 error page _(correct)_.
> <img alt="Screenshot 2023-11-01 at 12 40 14" width="948" src="https://user-images.githubusercontent.com/1035294/279663176-38053dd9-627f-4820-9cdb-e3db8f9de2d5.png">
> ...
> Unfortunately, this is not the cause of the problem I reported, because I already have `django.middleware.locale.LocaleMiddleware` in my [demo.zip](https://github.com/fabiocaccamo/django-admin-interface/files/13362696/demo.zip) `settings.MIDDLEWARE` list.

@merwok have you the possibility to double-check this?

julianwachholz commented:
In the meantime, I've come across a bug in the current version where it would redirect me to a URL like `/en-gb/en-gb/...`. This is fixed in my branch which I'm currently using in our production system.

fabiocaccamo commented:
@julianwachholz how does it happen? Could you submit a PR with test and bugfix for this too?

julianwachholz commented:
I haven't been able to reproduce it with my changes. Fixing it in `main` doesn't seem useful to me if it would be overridden with the changes from this PR?

fabiocaccamo commented:
Can't you fix it in this PR?

julianwachholz commented:
It is fixed in this PR

merwok commented:
Sorry I could not test or answer when you asked Fabio – my project all use LocalMiddleware so I wasn’t in a state to reproduce the problem noted.

fabiocaccamo commented:
@merwok no problem, don't worry.
